eventsource/eventsource-cross-origin-preflight.window.js

```diff
@@ -0,0 +1,35 @@
+// META: title=EventSource: cross-origin preflight
+// META: script=/common/utils.js
+
+const crossdomain = location.href.replace('://', '://élève.').replace(/\/[^\/]*$/, '/');
+const origin = location.origin.replace('://', '://xn--lve-6lad.');
+
+[
+  ['safe `last-event-id` (no preflight)', 'safe'],
+  ['unsafe `last-event-id` (too long)', 'long'],
+  ['unsafe `last-event-id` (unsafe characters)', 'unsafe']
+].forEach(([name, fixture]) => {
+  async_test(document.title + ' - ' + name).step(function () {
+    const uuid = token();
+    const url = crossdomain + 'resources/cors-unsafe-last-event-id.py?fixture=' + fixture + '&token=' + uuid;
+
+    const source = new EventSource(url);
+
+    // Make sure to close the EventSource after the test is done.
+    this.add_cleanup(() => source.close());
+
+    // 1. Event will be a `message` with `id` set to a CORS-safe value, then disconnects.
+    source.addEventListener(
+      'message',
+      this.step_func((evt) => assert_equals(evt.data, fixture)),
+    );
+
+    // 2. Will emit either `success` or `failure` event. We expect `success`,
+    //    which is the case if `last-event-id` is set to the same value as received above,
+    //    and a preflight request has been sent for the unsafe `last-event-id` headers.
```

Comment:
Wait shouldn't this be `open` and `error` events?

Reply:
Hmm, the way I read the spec/ask here is that if the `last-event-id` header contains any "unsafe bytes", AND the request is going cross-origin, then we should do a preflight request that contains `access-control-request-headers: last-event-id`. If the server then responds with `access-control-allow-headers: last-event-id`, we are okay to send the "unsafe" value to the server. If the server does not respond to the preflight request with that header, it will fail the request. Can you confirm if this is the intended behavior, and if it is not correct, outline what the correct behavior is?

As for the use of `success`/`failure`: We could theoretically send back a 400 or something if the preflight wasn't run, but the EventSource API purposely gives very little information on errors (basically nothing at all). If we use the generic error event, it's hard/harder to then detect _why_ the request failed. I prefer to handle things more explicitly - by sending [named events](https://github.com/rexxars/wpt/blob/fetch-cors-safelist-last-event-id/eventsource/resources/cors-unsafe-last-event-id.py#L94-L96) for success/failure we'll have access to the data field and can use that in the error output, eg "expected preflight, did not get one" or "got <unexpected last-event-id>, expected <expected last-event-id>". I find it both helps implementers figure out what is going wrong, as well as anyone reading the text/fixture, as the different cases are more explicitly called out.